### PR TITLE
[Caching] Ensure cache cleared when rector.php changed on FileHashComputer::compute()

### PR DESCRIPTION
**Pull request by samsonasik**
@TomasVotruba @staabm I found that when `rector.php` changed, the cache doesn't cleared and requires `--clear-cache` to apply change **when add/remove rule from the list directly (one by one)**.

```php
$rectorConfig->rules([
       // add new rule into the list...
]);
```

`SimpleParameterProvider::hash()` seems only cover when `parameters` (paths, skip, bootstrap) changed, not when rule added/removed from the list.

This patch ensure cache cleared when `rector.php` changed by read the `rector.php` content to be part of content passed to `sha1()`.

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

File: packages/Caching/Config/FileHashComputer.php [hunk 1]
```diff
@@ -17,7 +18,7 @@ public function compute(string $filePath): string
         $this->ensureIsPhp($filePath);
 
         $parametersHash = SimpleParameterProvider::hash();
-        return sha1($filePath . $parametersHash);
+        return sha1(FileSystem::read($filePath) . $parametersHash);
```
Comment (samsonasik) [inline packages/Caching/Config/FileHashComputer.php]: another solution is by using `sha1_file()` instead:

```suggestion
        return sha1_file($filePath);
```

when `rector.php` changed, even only space, the hash also changed:

```bash
➜  rector-src git:(ensure-always-refresh) php -r 'echo sha1_file(__DIR__ . "/rector.php");';
71902587bce991de196cf2c1f79f1083e3834bbf%      

// change the file ...
// re-run
                                                                                                                                 
➜  rector-src git:(ensure-always-refresh) php -r 'echo sha1_file(__DIR__ . "/rector.php");';
59b7d5b0e9d9672186774efb691a9ccc8dd5ebac% 
```

Comment (samsonasik) [reply packages/Caching/Config/FileHashComputer.php]: @TomasVotruba @staabm what do you think?

Comment (samsonasik) [reply packages/Caching/Config/FileHashComputer.php]: also, `sets` is not part of `$parametersHash`, so only rely on  `$parametersHash`, change `$rectorConfig->sets()` content won't change the cached files, so definitely requires 2 options:

- read the file + parameter hash like current proposed solution
- use sha1_file() instead.

on possible improvement: when using sets, I think the existing rules inside sets need to be extracted + hashed, then compare so when new rule added into set, the cache automatically cleared, but that may need a different PR :)

Comment (staabm) [reply packages/Caching/Config/FileHashComputer.php]: > sha1_file() instead

👍

---


What comes to mind: should we also support auto cache clear when contents of imported configs change?

---

Is the rector version itself also a criteria to invalidate the cache?

Comment (samsonasik) [reply packages/Caching/Config/FileHashComputer.php]: as I commented previously, for imported config changed, eg: via sets() method, that should be in separate PR.

Rector version changed should be also in separate PR to ease review and revert.

**Comment by staabm** [comment]
A new test would be awesome

File: packages/Caching/Config/FileHashComputer.php (hunk 1 above)
Comment (samsonasik) [reply packages/Caching/Config/FileHashComputer.php]: I updated to use `sha1_file()` https://github.com/rectorphp/rector-src/pull/4899/commits/7a3b7b34d91b5095cb137790720ab3473cc74f18

File: packages/Caching/Config/FileHashComputer.php [hunk 2]
```diff
@@ -30,7 +27,7 @@ private function ensureIsPhp(string $filePath): void
 
         throw new ShouldNotHappenException(sprintf(
             // getRealPath() cannot be used, as it breaks in phar
-            'Provide only PHP file, ready for Symfony Dependency Injection. "%s" given',
+            'Provide only PHP file, ready for Laravel Dependency Injection. "%s" given',
```
Comment (staabm) [inline packages/Caching/Config/FileHashComputer.php]:
```suggestion
            'Provide only PHP file, ready for Dependency Injection. "%s" given',
```

File: src/Configuration/Parameter/SimpleParameterProvider.php [hunk 3]
```diff
@@ -94,16 +94,6 @@ public static function provideBoolParameter(string $name, ?bool $default = null)
...
 
-    /**
-     * @api
-     * For cache invalidation
-     */
-    public static function hash(): string
-    {
-        $parameterKeys = self::$parameters;
-        return sha1(serialize($parameterKeys));
-    }
```
Comment (samsonasik) [inline src/Configuration/Parameter/SimpleParameterProvider.php:105]: this method no longer needed 👍

**Comment by samsonasik** [comment]
@staabm I've added unit test for it https://github.com/rectorphp/rector-src/pull/4899/commits/71b7c5a9f54cd1e0646bdafdc3d298bcbea0751c

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready 👍

**Comment by samsonasik** [comment]
@harikt by this PR, your old issue should finally resolved :)

- https://github.com/rectorphp/rector/issues/6996

**Comment by samsonasik** [comment]
It possibly already resolved at PR:.

- https://github.com/rectorphp/rector-src/pull/4902

I am closing for now, I will check when back to PC.

Next possible improvement is include Rector version in hash string so it will auto-clear after composer up got new version.

**Comment by staabm** [comment]
At least the testcase should be merged

**Comment by samsonasik** [comment]
@TomasVotruba I just tried your change on PR:

- https://github.com/rectorphp/rector-src/pull/4902

and unit test got failure:

```
There was 1 failure:

1) Rector\Tests\Caching\Config\FileHashComputer\FileHashComputerTest::testRectorPhpChanged
Failed asserting that two strings are not identical.

/Users/samsonasik/www/rector-src/packages-tests/Caching/Config/FileHashComputer/FileHashComputerTest.php:28
```

@TomasVotruba it seems `$rectorConfig->rules()` changed doesn't change the hash.

**Comment by samsonasik** [comment]
@staabm here the failing test case PR:

- https://github.com/rectorphp/rector-src/pull/4903

**Comment by harikt** [comment]
@samsonasik Happy to see the bug has been resolved.

**Pull request closed without merging**